File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/features/signingKeys/revokeSigningKeys/SigningKeyRevoker.java
package br.com.deltaglobalbank.identity.features.signingKeys.revokeSigningKeys;

import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.token.CannotRevokeActiveKeyException;
import br.com.deltaglobalbank.identity.domain.token.SigningKey;
import br.com.deltaglobalbank.identity.domain.token.SigningKeyNotFoundException;
import br.com.deltaglobalbank.identity.domain.token.SigningKeyRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class SigningKeyRevoker {

    private final SigningKeyRepository signingKeyRepository;

    public SigningKeyRevoker(SigningKeyRepository signingKeyRepository) {
        this.signingKeyRepository = signingKeyRepository;
    }

    @Transactional
    public boolean revoke(UUID id) {
        SigningKey key = signingKeyRepository.findById(id);
        if (key == null) {
            throw new SigningKeyNotFoundException();
        }

        return switch (key.status()) {
            case ACTIVE -> throw new CannotRevokeActiveKeyException();
            case REVOKED -> false;
            case RETIRED -> {
                key.revoke();
                signingKeyRepository.save(key);
                yield true;
            }
        };
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/features/signingKeys/revokeSigningKeys/SigningKeyRevoker.kt
package br.com.deltaglobalbank.identity.features.signingKeys.revokeSigningKeys

import br.com.deltaglobalbank.identity.domain.token.CannotRevokeActiveKeyException
import br.com.deltaglobalbank.identity.domain.token.SigningKeyNotFoundException
import br.com.deltaglobalbank.identity.domain.token.SigningKeyRepository
import br.com.deltaglobalbank.identity.domain.token.SigningKeyStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class SigningKeyRevoker(
    private val signingKeyRepository: SigningKeyRepository
) {
    @Transactional
    fun revoke(id: UUID): Boolean {
        val key = signingKeyRepository.findById(id)
            ?: throw SigningKeyNotFoundException()

        return when (key.status()) {
            SigningKeyStatus.ACTIVE  -> throw CannotRevokeActiveKeyException()
            SigningKeyStatus.REVOKED -> false
            SigningKeyStatus.RETIRED -> {
                key.revoke()
                signingKeyRepository.save(key)
                true
            }
        }
    }
}